File: data-structures/doubly-linked-list/doubly-linked-list.go
package main

import (
	"fmt"
)

type Node struct {
	value    int32
	next     *Node
	previous *Node
}

type DoublyLinkedList struct {
	head   *Node
	tail   *Node
	length int32
}

//pushes value at the end of the linked list
func (dl *DoublyLinkedList) push(value int32) {
	node := Node{value, nil, nil}

	if dl.head == nil {
		dl.head = &node
		dl.tail = &node
	} else {
		dl.tail.next = &node
		node.previous = dl.tail
		dl.tail = &node
	}
	dl.length++
}

//traverse and print the list
func (dl *DoublyLinkedList) traverse() {
	current := dl.head
	for current != nil {
		fmt.Println(current.value)
		current = current.next
	}
}

//push new value at the start of the list
func (dl *DoublyLinkedList) unshift(value int32) *DoublyLinkedList {
	node := Node{value, nil, nil}
	if dl.head == nil {
		dl.head = &node
		dl.tail = &node
	} else {
		node.next = dl.head
		dl.head.previous = &node
		dl.head = &node
	}
	dl.length++
	return dl
}

//pop last element
func (dl *DoublyLinkedList) pop() *Node {
	if dl.length == 0 {
		return nil
	}
	removedNode := dl.head
	if dl.length == 1 {
		dl.head = nil
		dl.tail = nil
		dl.length--
		return removedNode
	} else {
		dl.tail = dl.tail.previous
		dl.tail.next = nil
		removedNode.previous = nil
		dl.length--
		return removedNode
	}
}

// remove first element from the list
func (dl *DoublyLinkedList) shift() *Node {
	head := dl.head
	if head == nil {
		return nil
	}
	if dl.length == 1 {
		dl.head = nil
		dl.tail = nil
	} else {
		dl.head.next.previous = nil
		dl.head = dl.head.next
		head.next = nil
	}
	dl.length--
	return head
}

//get element at the position
func (dl *DoublyLinkedList) get(position int32) *Node {
	if dl.length == 0 || position >= dl.length {
		return nil
	}
	var currentNode *Node
	if position <= dl.length/2 {
		currentNode = dl.head
		for i := 0; int32(i) < dl.length; i++ {
			if position == int32(i) {
				return currentNode
			}
			currentNode = currentNode.next
		}
	} else {
		currentNode = dl.tail
		for i := dl.length - 1; int32(i) >= 0; i-- {
			if position == int32(i) {
				return currentNode
			}
			currentNode = currentNode.previous
		}
	}
	return currentNode
}

//accepts a position and a value and if the node is present at that position. the new value is set at that position.
func (dl *DoublyLinkedList) set(position int32, value int32) bool {
	node := dl.get(position)
	if node == nil {
		return false
	}
	node.value = value
	return true
}

//accepts a position and a value and if the node is present at that position. the new node inserted at that position.
func (dl *DoublyLinkedList) insert(position int32, value int32) bool {
	if dl.length == 0 || position >= dl.length {
		return false
	}
	if position == 0 {
		dl.unshift(value)
		return true
	} else if position == dl.length-1 {
		dl.push(value)
		return true
	} else {
		node := dl.get(position)
		newNode := Node{value, nil, nil}
		newNode.previous = node.previous
		node.previous.next = &newNode
		node.previous = &newNode
		newNode.next = node
		dl.length++
		return true
	}
}
func (dl *DoublyLinkedList) remove(position int32) *Node {
	if dl.length == 0 || position >= dl.length {
		return nil
	}
	if position == 0 {
		return dl.shift()
	} else if position == dl.length-1 {
		return dl.pop()
	} else {
		node := dl.get(position)
		node.previous.next = node.next
		node.next.previous = node.previous
		node.next = nil
		node.previous = nil
		dl.length--
		return node

	}
}

func main() {
	dl := DoublyLinkedList{}
	dl.push(1)
	dl.push(2)
	dl.push(3)
	dl.push(4)

	dl.remove(0)
	dl.traverse()

}

File: data-structures/doubly-linked-list/doubly-linked-list.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
        this.previous = null;
    }

}

class DoublyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) {
        var node = new Node(value);
        if (this.length === 0) {
            this.head = node;
            this.tail = node;

        } else {
            this.tail.next = node;
            node.previous = this.tail;
            this.tail = node;
        }
        this.length++
        return this
    }

    pop() {
        if (this.length === 0) {
            return null;
        }
       
        var removedNode = this.tail;
        if (this.length === 1) {
            this.head = null
            this.tail = null
            this.length--;

            return removedNode;
        } else {
           
            this.tail = this.tail.previous;
            this.tail.next = null;
            removedNode.previous = null;
            this.length--;

            return removedNode;

        }

    }

    shift() {
        if (this.length === 0) {
            return null;
        }
        var removedNode = this.head;
        if (this.length === 1) {
            this.head = null
            this.tail = null
        } else {
            this.head.next.previous = null
            this.head = this.head.next;
            removedNode.next = null
        }
        this.length--;
        return removedNode;
    }

    unshift(value) {
        var node = new Node(value);
        if (!this.head) {
            this.head = node;
            this.tail = node;
        } else {
            node.next = this.head;
            this.head.previous = node;
            this.head = node;
        }
        this.length++;
        return this;
    }

    get(position) {
        if (position < 0 || position >= this.length) {
            return null
        }
        var currentNode
        if (position <= this.length / 2) {
            currentNode = this.head;
            for (var i = 0; i < this.length; i++) {
                if (position === i) {
                    return currentNode;
                }
                currentNode = currentNode.next;
            }
        } else {
            currentNode = this.tail;
            for (var i = this.length - 1; i >= 0; i--) {
                if (position === i) {
                    return currentNode;
                }
                currentNode = currentNode.previous;
            }

        }
        return currentNode;

    }

    set(position, value) {
        var node = this.get(position);
        if (!node) {
            return false
        }
        node.value = value;
        return true;

    }

    insert(position, value) {
        if (this.length === 0 || position >= this.length) {
            return false
        }
        if (position === 0) {
            this.unshift(value);
            return true;
        } else if (position === this.length - 1) {
            this.push(value)
            return true;

        }
        var node = this.get(position);
        var newNode = new Node(value);
        newNode.previous = node.previous;
        node.previous.next = newNode;
        node.previous = newNode;
        newNode.next = node;
        this.length++;
        return true;
    }

    remove(position) {
        if (this.length === 0 || position >= this.length) {
            return false
        }
        if (position === 0) {
            return this.shift();

        } else if (position === this.length - 1) {
            return this.pop()

        }

        var node = this.get(position);
        node.previous.next = node.next;
        node.next.previous = node.previous;
        node.next = null
        node.previous = null
        this.length--;
        return node;
    }

    traverse() {
        var currentNode = this.head;
        if (!currentNode) {
            return null
        }
        for (var i = 0; i < this.length; i++) {
            console.log(currentNode.value);
            currentNode = currentNode.next;
        }
    }

}

var dll = new DoublyLinkedList();
